### DISCUSSION TITLE
Functionality to assist in Amateur Radio Direction Finding (ARDF)

### DISCUSSION THREAD

**Discussion by darksidelemm** [Tunas1337/UV-K5-Modded-Firmwares#46] [Ideas]
Based on @Tunas1337 's comment in [this issue](https://github.com/piotr022/UV_K5_playground/issues/22#issuecomment-1637470396) I thought I'd make some notes on what functionality would be handy to assist with Amateur Radio Direction Finding.

### Background
I've been involved in fox-hunting for many years, and have picked up a few things over that time as to what kind of tools are useful to help in this activity.

### What you do while Foxhunting
The aim of foxhunting is to find a hidden transmitter. It might be a few hundred metres away, it might be many kilometers away. The signal strengths your receiver might see over a foxhunt might vary from being right on the threshold of detectability, to massively strong result in saturation of your receiver

The transmitter is *usually* transmitting fairly regularly (often not continuously), and you usually (hopefully?) have some kind of directional antenna. Let's not get into doppler systems and phased arrays in this discussion :-)

Throughout the hunt, you are trying to figure out the direction to the transmitter. You can do this in a few days, but generally it's by swinging a directional antenna around and trying to figure out the direction of strongest signal. Head in that direction, take another measurement, repeat a few (many?) times and hopefully you've found the transmitter.

### Directional Antennas and Power Measurement
Directional antennas used in foxhunts can vary quite a bit. You can have antennas which have reasonable gain and narrow beam-widths on the higher bands (like 70cm, where a 7-element beam is quite practical), or antennas which might have a fairly wide beam, like the typical 3-element yagis used on the 2m band.

Looking at the 2m case, a typical 3-element yagi might have a 3dB beam-width as large as 60 degrees. This means that the signal level is only going to vary by 3dB when moving the antenna across 60 degrees. This is not much! It's almost impossible to view on a regular s-meter (ideally 6dB per s-point). Being able to view the actual RSSI number helps a LOT here. 

### The other equipment I use
One of the best tools that I have for fox-hunting is a ['VK3YNG Sniffer'](http://www.foxhunt.com.au/2m_sniffer/manual.htm). It's essentially an audible power meter, and converts power level to tone-pitch. To handle the wide range of signal levels it will encounter over a hunt it has a neat trick it uses:
* The sniffer has 10 'ranges', which are essentially different attenuation steps. Each range is about ~15 dB wide, with 3-5dB overlap (unsure exactly what the overlap is). As the signal moves into the edges of the overlap region, the receiver steps into the next (higher or lower) range. The current range (0 - 9) is indicated on a display.
* Within each range, the power-to-tone conversion is represented logarithmically, from about 100 Hz (0dB in the current range) up to 6 kHz (~15 dB in the range). This makes it really easy to 'hear' the variation in power as you swing the beam, even with fairly small changes in power (e.g. over that 3dB beamwidth).

So, as the you get closer, you see the range number increase, but you still get that variation in tone pitch to get a direction!

Some videos of these in action:
* https://photos.app.goo.gl/z6kpef6RLpgo7EGi8
* https://youtu.be/88Spk-OV1Vw?t=383 (In this case the fox is transmitting short pulses)

### UV-K5 Functionality for Fox-hunting

* RSSI measurement!
  * The display should show signal level in as high resolution as is available from the radio IC (in this case 0.5 dB), with a decent update rate and as low latency as possible.
  * The display should be always running, even if the squelch is not broken.
  * The display should be big :-) When foxhunting I don't really care about that second VFO. I just want to see a big signal strength indication and not have to squint!

* RSSI measurement linearity
  * Ideally the RSSI value should be linear over a wide range of powers. What we don't want is for it to clip or compress when the signal levels get too high (so, being able to add attenuation would be nice!). 
  * Conversely, we want to be able to detect weak signals too (This one is a bit harder, you will be running up against the limitations of the receiver at this point).

* RSSI Display / Representation ideas - Some extra thoughts on RSSI display and how to show it in an easy-to-interpret manner.
  * Just the RSSI number (with 0.5 dB resolution) in big text will be a good start! This is better than the current S-meter.
  * A S-meter bar-graph which shows the signal level within a 'range' (like the VK3YNG sniffer) would be an interesting next step. 
  * Power to Tone pitch? I suspect the resolution of the RSSI measurement might make this not work so well unfortunately.

**Comment by Tunas1337** [comment]
I think this is a splendid idea!

One thing I'd add is that we can already plot a graph for the RSSI, and so it would be very nice to have maybe a visual indication over ~10 seconds. Converting RSSI to pitch and using the built in tone generator would be possible, but the tone generation blocks input and other tasks (as is evident if you have menu beeps on) so that might impede us.

**Comment by darksidelemm** [comment]
I'd certainly consider the tone representation stuff as a 'stretch goal'. If you're having issues with blocking when generating tones then it's probably not going to work.
I like the idea of the visual indication over 10 seconds, but it would need that 'stepping' approach to be able to see the change in level as you swing a beam around - remember that when we're trying to get an accurate direction we might only be seeing very small variations in receive power.

**Comment by darksidelemm** [comment]
I had the opportunity to see how the RSSI values responded to known signal levels from a calibrated-ish signal generator. The results are here: https://docs.google.com/spreadsheets/d/1dKfjRbv43sim1P9o6_B9eWVQTT_qL62H/edit?usp=sharing&ouid=104299818319616713136&rtpof=true&sd=true

A few notes:
- The linear region of the RSSI meter is quite large, and stretches from -118 dBm up to -15 dBm
- I think I can see evidence of the AGC in the chip kicking in around -15 dBm. 
- At very high input levels ~-3 dBm I think there's some kind of rollover going on.

Being able to override the AGC would be an interesting addition to this, to essentially force in additional attenuation when you're close to the transmitter.

**Comment by Tunas1337** [comment]
The AGC seems to be somewhat programmable, although your guess is as good as mine what the arbitrary values mean. We have limited documentation about the chip, and I'm grateful we have even that much :P

In short, once we know what we're doing, that's definitely possible.

**Comment by darksidelemm** [comment]
I had a go at using the RSSI indication on a foxhunt today. While it's fine for hunting foxes transmitting constant power, we have a lot of shorter-range foxes that pulse on and off.
The RSSI indication updates so quickly that the indication is almost unreadable with these kind of foxes. Video here: https://photos.app.goo.gl/7YB2yUf3t2iALTJf9

To be able to hunt these, you'd need to do some kind of running average on the RSSI data.

**Reply by Tunas1337** [reply]
Have you considered, instead, using the RSSI_runner that shows a little running plot of the RSSI over time (10 seconds, if I'm not mistaken?)

**Reply by darksidelemm** [reply]
The plot is *far* too compressed in dynamic range to be useful for this. Remember that the 3dB beamwidth of the antennas we are using is 60 degrees or so. You can't even see 3dB of signal strength change on the RSSI plot.
You'd need to implement the 'ranging' approach used in the YNG sniffer to be able to see anything.

In the following image i'd just swung the beam across 180 degrees, past one of these transmitters. You can barely tell anything has changed. 
![image](https://github.com/Tunas1337/UV-K5-Modded-Firmwares/assets/1040275/c9319884-1965-47fd-9e9c-09ba72e9392c)

**Reply by Tunas1337** [reply]
Sorry, early morning, forgot the context. Yes, you're absolutely right.
Would it help if it was a full-screen graph, *with* averaging?

**Reply by darksidelemm** [reply]
It would probably only be a marginal improvement, especially if its still showing the full dynamic range of the RSSI measurement.

The RSSI text output is fine in most cases, its just in this situation (pulsed transmitter), it needs some running averaging (or a running peak-hold) to be useful.
